presto-main/src/main/java/com/facebook/presto/operator/TableWriterOperator.java

```diff
@@ -193,15 +217,21 @@ public void addInput(Page page)
         requireNonNull(page, "page is null");
         checkState(needsInput(), "Operator does not need input");
 
-        Block[] blocks = new Block[inputChannels.size()];
-        for (int outputChannel = 0; outputChannel < inputChannels.size(); outputChannel++) {
-            blocks[outputChannel] = page.getBlock(inputChannels.get(outputChannel));
+        Block[] blocks = new Block[columnChannels.size()];
+        for (int outputChannel = 0; outputChannel < columnChannels.size(); outputChannel++) {
+            blocks[outputChannel] = page.getBlock(columnChannels.get(outputChannel));
         }
 
+        statisticAggregation.addInput(page);
+        ListenableFuture<?> blockedOnAggregation = statisticAggregation.isBlocked();
         CompletableFuture<?> future = pageSink.appendPage(new Page(blocks));
         updateMemoryUsage();
-        if (!future.isDone()) {
-            this.blocked = toListenableFuture(future);
+        ListenableFuture<?> blockedOnWrite = toListenableFuture(future);
+        if (blockedOnWrite.isDone() && blockedOnAggregation.isDone()) {
+            blocked = NOT_BLOCKED;
+        }
+        else {
+            blocked = allAsList(blockedOnAggregation, blockedOnWrite);
```

Comment:
remove `if`, `else`'s code is generic enough.

if you don't want to go through `com.google.common.util.concurrent.Futures#allAsList(com.google.common.util.concurrent.ListenableFuture<? extends V>...)` in all-done case, create a helper method that does the `if`